🐾 Support a new details item field for the remote provider's web ui

sgratch opened the pull request:
Reference: https://github.com/kubev2v/forklift-console-plugin/issues/1106

Enhance the providers details page by supporting a new details item field for showing/editing/clicking the provider's web ui url.

This field is special since it doesn't include a title (appears beneath the provider's name item field) and has 2 modes:
1. If empty, a proper "empty field" text is displayed
2. If non empty, an external hyper link is displayed.

The field's value is calculated as follows:
1. Search for an annotation `forklift.konveyor.io/providerUI`.
2. If no annotation is set, try to auto calculate the url, based on provider's api server url.
3. If not 1, 2 then display an "empty field" text.

Once edited, the edited value will be saved as an annotation.


### Screenshots (examples for the RHV provider):
![Screenshot from 2024-05-02 19-46-57](https://github.com/kubev2v/forklift-console-plugin/assets/18169498/361256b0-51f8-4ddb-a8b4-a3c43b3b62e8)
![Screenshot from 2024-05-12 22-30-08](https://github.com/kubev2v/forklift-console-plugin/assets/18169498/72224178-9374-4699-a0ae-eb23410c21ee)
![Screenshot from 2024-05-12 22-33-57](https://github.com/kubev2v/forklift-console-plugin/assets/18169498/1b335bee-c2b6-48ac-985e-e04f54ee5252)

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/UIDetailsItem.tsx:
```diff
@@ -0,0 +1,59 @@
+const getProviderUIAnnotation = (provider: V1beta1Provider): string =>
+  provider?.metadata?.annotations?.['forklift.konveyor.io/providerUI'];
+
+export const UIDetailsItem: React.FC<UIDetailsItemProps> = ({
```
it's an 
```ts
<EditableContentButton content={content} onEdit={onEdit} />
```

not an detailsitem

sgratch replied:
I understand your point. Since this component is presented within the providers details section only  so I do want to indicate that it's a details field and not some general field.

All fields within the provider details section are named as the content they include, e.g.:
TypeDetailsItem
NamespaceDetailsItem
URLDetailsItem
CredentialsDetailsItem
OwnerDetailsItem

Part of them are editable, part are not, few display a text and few display a link.
The only uniqueness of this `UIDetailsItem` comp is that it doesn't include a title. 
And  I do agree with your follow up comment regarding the "UI" naming part.

So maybe we can rename it to `WebUIDetailsItemNoTitle` or something like that.

yaacov replied:
I whould like it to be part of the Name DetailItem, the Name details item will have 2 content items, can you make it part of the Name details item content ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/UIDetailsItem.tsx:
```diff
@@ -0,0 +1,59 @@
+    getProviderUIAnnotation(provider) ?? (calcProviderUI ? calcProviderUI(provider) : null);
+
+  return (
+    <DetailsItem
```
```suggestion
    <EditableContentButton
```

sgratch replied:
I prefer the current implementation of wrapping that with the `DetailsItem` for 2 reasons:
1. it's much more flexibility for  future usages. We might re-use this with a title, disable the editable option etc. 
Current implementation enable all options for items within the details page.

2. Additionally, if the user doesn't have patching permissions, then it will be `NonEditableContent` and not `EditableContentButton`
So we can move this logic to `UIDetailsItem` but why duplicating that?

yaacov replied:
My concern it that it is a content not an item, I don't want to use the name "Item" for a content

> t's much more flexibility 

It is a content, it is flexibel enogh

> if the user doesn't have patching permissions,

That is a good point, I prefer to address this concern by enhancing  `EditableContentButton` so it will render `NonEditableContent` when `onEdit` is falsy, that way it is still a content item, but it have more general use case, it will also simplify  `DetailsItem`

sgratch replied:
The `UIDetailsItem` was replaced with adding the web ui link into the `NameAndWebUiLinkDetailsItem` comp.
So this is not relevant anymore.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/components/DetailsPage/DetailItem.tsx:
```diff
@@ -37,18 +37,16 @@ export const DetailsItem: React.FC<DetailsItemProps> = ({
     <DescriptionListGroup>
-      {helpContent ? (
-        <DescriptionTitleWithHelp
+      {title ? (
```
see comment https://github.com/kubev2v/forklift-console-plugin/pull/1135/files#r1588064079

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/modals/EditProviderUI/EditProviderUIModal.tsx:
```diff
@@ -0,0 +1,29 @@
+    case 'openstack':
+    case 'vsphere':
+    default:
+      return <></>;
```
not yet implemented, what am i missing ?

p.s.
if the modal is not yet implemented, make this PR a WIP, and implement it in a different PR, once the modal merged, mark this one as ready for review.

yaacov replied:
ahh, I see this is an infrastructure PR, no changes in the UI,

OK :+1:

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/ProviderDetailsItem.tsx:
```diff
@@ -2,6 +2,16 @@ import { ReactNode } from 'react';
 import { V1beta1Provider } from '@kubev2v/types';
 
+/**
+ * Type for the props of the ProviderDetailsItemProps component.
```
Nice !

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/components/DetailsPage/DetailItem.tsx:
```diff
@@ -175,7 +199,7 @@ export const NonEditableContent: React.FC<{ content: ReactNode }> = ({ content }
  */
 export type DetailsItemProps = {
-  title: string;
+  title?: string;
```
see comment https://github.com/kubev2v/forklift-console-plugin/pull/1135/files#r1588064079

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/UIDetailsItem.tsx:
```diff
@@ -0,0 +1,59 @@
+      onEdit={
+        canPatch &&
+        provider?.metadata &&
+        (() => showModal(<EditProviderUIModal resource={provider} content={content} />))
```
```suggestion
        (() => undefined)
```
this is not implemented yet ...

yaacov replied:
and then you don't need to add a null modal ...

sgratch replied:
I don't want to add it for the 1st  provider's implementation  if I already prepare this infra PR then why not preparing it here so that for each provider, I'll only touch the `EditProviderUIModal`  implementation. And won't touch this component.
But it's not important. I can also add it for the 1st provider.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/UIDetailsItem.tsx:
```diff
@@ -0,0 +1,59 @@
+  resource: provider,
+  canPatch,
+  text,
+  calcProviderUI,
```
i like the concept of passing a method here, i have concern about the name `ProviderUI` sounds like some UI thing, we are doing a "web UI" everything here is a UI the components are UI componenet the application is a UI application, etc ... , we need to be clear that UI here is in the context of a link to a web based management system vs the context of some other UI thing we are doing.

sgratch replied:
`calcProviderWebUIConsole` maybe? 
I didn't want to use long names, but we can think of something

yaacov replied:
I prefer somthing with "link" in it, so we know it's a URL link string for the console and not the "console", any idea ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/UIDetailsItem.tsx:
```diff
@@ -0,0 +1,59 @@
+          </ExternalLink>
+        ) : (
+          <span className="text-muted">
+            {t('Optional: Click the edit button for setting the provider web UI path')}
```
I'm ok with the text, my concern is that it's long, can we convey the same information in a shorter text ? do we need the "optional" text ?

sgratch replied:
`Click the pencil for setting provider's  web UI path` ?
Since we don't display a title, we must explain here what's this content for....

yaacov replied:
maybe link instead of "path", "link" indicate it is a link, while "path" may mean it's some internal path ?

sgratch commented:
All comments above were addressed.

The new design includes the following changes:
1. Replaced the provider's `NameDetailsItem` with `NameAndWebUiLinkDetailsItem`. 

3. For supporting a flexible number of content fields per a detail item, an array of **ReactNode** `contents`  prop and a corresponding array of `onEdits` prop was added such that:
    for every i:
        `contents`[i] is displayed and iff `onEdits`[i] callback exists then this `contents`[i] field is editable. 
        Otherwise it's non editable.

4. One limitation is that for avoiding changing all `DetailsItem` usages throughout the whole code (risky...) then I left the `content`, `onEdit` props as is for displaying  the 1st/only content field (backward compatible) and just added the new props of `contents`, `onEdits` for displaying the 2nd and forward fields.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/components/DetailsPage/DetailItem.tsx:
```diff
@@ -132,55 +162,53 @@ export const DescriptionTitle: React.FC<{ title: string }> = ({ title }) => (
   moreInfoLabel?: string;
   moreInfoLink?: string;
   crumbs?: string[];
+  content: ReactNode;
+  contents?: ReactNode[];
```
```suggestion
  content: ReactNode | ReactNode[];
```

and then when using it, ensure it'a an array before using it, for example by using a helper method:

``` ts
// Function to ensure the content is always an array of ReactNode
function ensureArray(content: ReactNode | ReactNode[]): ReactNode[] {
  if (Array.isArray(content)) {
    return content;
  } else {
    return [content]; // Wrap the single ReactNode in an array
  }
}
```

sgratch replied:
I already saw that this solution used by you in our code for the `TableLabelCell` labels:  
https://github.com/kubev2v/forklift-console-plugin/blob/e00a3ddf5b4adbe06f4c267b17b9d2f70cddc29f/packages/forklift-console-plugin/src/modules/Providers/utils/components/TableCell/TableLabelCell.tsx#L55-L57

I less liked this since dividing to a content and an array of contents implies that there should be at least one content for the `DetailsItem` instead of one array that can be null (unlike labels for the table). But you can still set `content` to `null` then it's anyway not really enforced.
OK, I'll change it

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/NameAndWebUiLinkDetailsItem.tsx:
```diff
@@ -0,0 +1,91 @@
+        helpContent={helpContent ?? defaultHelpContent}
+        crumbs={['Provider', 'metadata', 'name']}
+        // for name content field
+        content={nameContent}
+        // for web UI link content field
+        contents={[
+          webUILinkContent ? (
+            <ExternalLink text={webUILinkText} href={webUILinkContent} isInline={true}>
+              {webUILinkContent}
+            </ExternalLink>
+          ) : (
+            <span className="text-muted">
+              {canPatch && provider?.metadata
+                ? t('Click the pencil for setting provider web UI link')
+                : t('No value for provider web UI link')}
+            </span>
+          ),
+        ]}
```
```suggestion
        content={[nameContent, 
          webUILinkContent ? (
            <ExternalLink text={webUILinkText} href={webUILinkContent} isInline={true}>
              {webUILinkContent}
            </ExternalLink>
          ) : (
            <span className="text-muted">
              {canPatch && provider?.metadata
                ? t('Click the pencil for setting provider web UI link')
                : t('No value for provider web UI link')}
            </span>
          ),
        ]}
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/NameAndWebUiLinkDetailsItem.tsx:
```diff
@@ -0,0 +1,91 @@
+ * @component
+ * @param {DetailsItemProps} props - The props of the details item.
+ */
+export const NameAndWebUiLinkDetailsItem: React.FC<NameAndWebUiLinkDetailsItemProps> = ({
```
```suggestion
export const NameAndLinkDetailsItem: React.FC<NameAndWebUiLinkDetailsItemProps> = ({
```

sgratch replied:
The name is long but just `Link` is not self explained...if you insist on a shorter name then I prefer `NameUiLinkDetailsItem`

yaacov replied:
ok, I think the "and" is important, so `NameAndUiLinkDetailsItem` WDYT ?

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/NameAndWebUiLinkDetailsItem.tsx:
```diff
@@ -0,0 +1,91 @@
+  const defaultMoreInfoLink =
+    'https://kubernetes.io/docs/concepts/overview/working-with-objects/names';
+  const defaultHelpContent = t(
+    "This field includes the name and an optional provider's web UI link. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. Provider's web UI link is editable and its value is either taken from YAML or auto calculated.",
```
```suggestion
    "A client-provided string that refers to an object in a resource URL, such as /api/v1/pods/some-name.

Only one object of a given kind can have a given name at a time. However, if you delete the object, you can make a new object with the same name.

Use the link below to access the web-based user interface for virtual machine management. You can store the link to the management system in the Provider metadata annotations value under the 'forklift.konveyor.io/providerUI' annotation.
```

sgratch replied:
> "A client-provided string that refers to an object in a resource URL" 

?? 
not clear.

yaacov replied:
> not clear.

I'ts coppied from https://kubernetes.io/docs/concepts/overview/working-with-objects/names, yes, they are not that clear, but it's a text the k8s team docs think is good, I'm ok with adding some information it you think it's needed to make it more clear, this text has lots of space.

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/components/DetailsPage/DetailItem.tsx:
```diff
@@ -27,37 +27,67 @@ import Pencil from '@patternfly/react-icons/dist/esm/icons/pencil-alt-icon';
+        crumbs={crumbs}
+      />
+      <DescriptionListDescription>
+        <ContentField content={content} onEdit={onEdit} />
+
+        {contents?.map((value, index) => (
+          <ContentField
+            key={'content-field-' + index}
+            content={value}
+            onEdit={onEdits ? onEdits[index] : null}
+          />
+        ))}
```
see: https://github.com/kubev2v/forklift-console-plugin/pull/1135/files#r1597880185

convert content to array:
`contents = ensureArray(contnet)`

and then use it as array
```suggestion
        {contents?.map((value, index) => (
          <ContentField
            key={'content-field-' + index}
            content={value}
            onEdit={onEdits ? onEdits[index] : null}
          />
        ))}
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/components/DetailsPage/DetailItem.tsx:
```diff
@@ -132,55 +162,53 @@ export const DescriptionTitle: React.FC<{ title: string }> = ({ title }) => (
   crumbs?: string[];
+  content: ReactNode;
+  contents?: ReactNode[];
   onEdit?: () => void;
+  onEdits?: (() => void)[];
```
see comment https://github.com/kubev2v/forklift-console-plugin/pull/1135/files#r1597880185

```suggestion
  onEdit?: (() => void) | (() => void)[];
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/views/details/components/DetailsSection/components/NameAndUiLinkDetailsItem.tsx:
```diff
@@ -0,0 +1,102 @@
+  const nameContent = provider?.metadata?.name;
+
+  // Read the annotation for getting the web UI link and only if empty, try to auto calculate the provider web UI path
+  const webUILinkContent =
+    getProviderUIAnnotation(provider) ?? (calcWebUILink ? calcWebUILink(provider) : null);
```
```suggestion
  const webUILink =
    getProviderUIAnnotation(provider) ?? (calcWebUILink ? calcWebUILink(provider) : null);
    
  const webUILinkContent = <ExternalLink text={webUILinkText} href={webUILinkContent} isInline={true}>
              {webUILinkContent}
            </ExternalLink>;
            
  const webUILinkEmpty = <span className="text-muted">
              {canPatch && provider?.metadata
                ? t('Click the pencil for setting provider web UI link')
                : t('No value for provider web UI link')}
            </span>;
```

yaacov commented on packages/forklift-console-plugin/src/modules/Providers/utils/helpers/ensureArray.ts:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Function to ensure that the content is always an array of ReactNode.
+ */
+export const ensureReactNodeArray = (node: ReactNode | ReactNode[]): ReactNode[] => {
+  if (Array.isArray(node)) {
+    return node;
+  } else {
+    return [node]; // Wrap the single ReactNode in an array
+  }
+};
+
+/**
+ * Function to ensure that the content is always an array of functions.
+ */
+export const ensureFuncArray = (func: (() => void) | (() => void)[]): (() => void)[] => {
+  if (Array.isArray(func)) {
+    return func;
+  } else {
+    return [func]; // Wrap the single ReactNode in an array
+  }
+};
```
```suggestion
export const ensureArray = (node: unknown | unknown[]): unknown[] => {
  if (Array.isArray(node)) {
    return node;
  } else {
    return [node]; // Wrap the single ReactNode in an array
  }
};
```
